syntax = "proto3";
package store;

service Store {
  rpc CreateStore(CreateStoreRequest) returns (CreateStoreResponse) {}
  rpc ReadStore(ReadStoreRequest) returns (ReadStoreResponse) {}
  rpc UpdateStore(UpdateStoreRequest) returns (UpdateStoreResponse) {}
  rpc UpdateStoreByAddress(UpdateStoreByAddressRequest)
      returns (UpdateStoreByAddressResponse) {}
  rpc DeleteStore(DeleteStoreRequest) returns (DeleteStoreResponse) {}
  rpc DeleteStoreByAddress(DeleteStoreByAddressRequest)
      returns (DeleteStoreByAddressResponse) {}
  rpc ReadStoreList(ReadStoreListRequest) returns (ReadStoreListResponse) {}
  rpc ReadStoreByAddress(ReadStoreByAddressRequest)
      returns (ReadStoreByAddressResponse) {}
  //
  rpc AddBookToStore(AddBookToStoreRequest) returns (AddBookToStoreResponse) {}
  rpc RemoveBookFromStore(RemoveBookFromStoreRequest)
      returns (RemoveBookFromStoreResponse) {}
  rpc GetAmountOfSpecificBookFromStore(GetAmountOfSpecificBookFromStoreRequest)
      returns (GetAmountOfSpecificBookFromStoreResponse) {}
  //
  rpc AddVinylToStore(AddVinylToStoreRequest)
      returns (AddVinylToStoreResponse) {}
  rpc RemoveVinylFromStore(RemoveVinylFromStoreRequest)
      returns (RemoveVinylFromStoreResponse) {}
  rpc GetAmountOfSpecificVinylFromStore(
      GetAmountOfSpecificVinylFromStoreRequest)
      returns (GetAmountOfSpecificVinylFromStoreResponse) {}

  rpc AddToStockInfo(AddToStockInfoRequest) returns (AddToStockInfoResponse) {}
  rpc RemoveFromStockInfo(RemoveFromStockInfoRequest)
      returns (RemoveFromStockInfoResponse) {}
  rpc ReadStockInfo(ReadStockInfoRequest) returns (ReadStockInfoResponse) {}
  rpc IncrementReservedStockInfo(IncrementReservedStockInfoRequest)
      returns (IncrementReservedStockInfoResponse) {}
  rpc DecrementReservedStockInfo(DecrementReservedStockInfoRequest)
      returns (DecrementReservedStockInfoResponse) {}

  rpc ReturnItemStockInfo(ReturnItemStockInfoRequest)
      returns (ReturnItemStockInfoResponse) {}
}

message StoreObject {
  string address = 1;
  string phone_number = 2;
  string email = 3;
  string created_at = 4;
  string updated_at = 5;
}

// ######## Request ########
message CreateStoreRequest {
  string address = 1;
  string phone_number = 2;
  string email = 3;
}
message UpdateStoreRequest {
  int32 id = 1;
  string address = 2;
  string phone_number = 3;
  string email = 4;
}

message UpdateStoreByAddressRequest {
  string address_match = 1;
  string address_update = 2;
  string phone_number = 3;
  string email = 4;
}

message DeleteStoreRequest { int32 id = 1; }
message DeleteStoreByAddressRequest { string address = 1; }
message ReadStoreRequest { int32 id = 1; }
message ReadStoreByAddressRequest { string address = 1; }
message ReadStoreListRequest {}

message AddBookToStoreRequest {
  int32 store_id = 1;
  int32 book_id = 2;
}
message RemoveBookFromStoreRequest {
  int32 store_id = 1;
  int32 book_id = 2;
}
message GetAmountOfSpecificBookFromStoreRequest {
  int32 store_id = 1;
  int32 book_id = 2;
}

message AddVinylToStoreRequest {
  int32 store_id = 1;
  int32 vinyl_id = 2;
}
message RemoveVinylFromStoreRequest {
  int32 store_id = 1;
  int32 vinyl_id = 2;
}
message GetAmountOfSpecificVinylFromStoreRequest {
  int32 store_id = 1;
  int32 vinyl_id = 2;
}

message AddToStockInfoRequest {
  string uuid = 1;
  int32 store_id = 2;
  int32 total_count = 3;
  int32 in_stock = 4;
  int32 reserved_count = 5;
}
message RemoveFromStockInfoRequest {
  string uuid = 1;
  int32 store_id = 2;
}
message ReadStockInfoRequest {
  string uuid = 1;
  int32 store_id = 2;
}
message IncrementReservedStockInfoRequest {
  string uuid = 1;
  int32 store_id = 2;
}
message DecrementReservedStockInfoRequest {
  string uuid = 1;
  int32 store_id = 2;
}

message ReturnItemStockInfoRequest {
  string uuid = 1;
  int32 store_id = 2;
}

// ######## Response #######

message CreateStoreResponse { string msg = 1; }
message ReadStoreResponse { StoreObject store = 1; }
message UpdateStoreResponse { string msg = 1; }
message UpdateStoreByAddressResponse { string msg = 1; }
message DeleteStoreResponse { string msg = 1; }
message DeleteStoreByAddressResponse { string msg = 1; }
message ReadStoreListResponse { repeated StoreObject store_list = 1; }
message ReadStoreByAddressResponse { StoreObject store = 1; }

message AddBookToStoreResponse { string msg = 1; }
message RemoveBookFromStoreResponse { string msg = 1; }
message GetAmountOfSpecificBookFromStoreResponse { int32 amount = 1; }

message AddVinylToStoreResponse { string msg = 1; }
message RemoveVinylFromStoreResponse { string msg = 1; }
message GetAmountOfSpecificVinylFromStoreResponse { int32 amount = 1; }

message AddToStockInfoResponse { string msg = 1; }
message RemoveFromStockInfoResponse { string msg = 1; }
message ReadStockInfoResponse {
  string uuid = 1;
  int32 store_id = 2;
  int32 total_count = 3;
  int32 in_stock = 4;
  int32 reserved_count = 5;
  string created_at = 6;
  string updated_at = 7;
}
message IncrementReservedStockInfoResponse { string msg = 1; }
message DecrementReservedStockInfoResponse { string msg = 1; }
message ReturnItemStockInfoResponse { string msg = 1; }
